### PR TITLE
Change return type of Raft::new() to Result

### LINKED ISSUE #94: Raft::new() returns a Result
Depends on:
* #88 

Currently calling `Raft::new()` will panic if it is provided an invalid configuration.

https://github.com/pingcap/raft-rs/blob/18f1cd8bb1a452dd9a2e3a8c0b507fc138c1862d/src/raft.rs#L390-L391

The consequence of this is that the user of the `Raft` is not provided an opportunity to handle certain types of error.

Fixing this issue would include changing `Raft::new()` to return a `Result<Raft<T>>` instead of a `Raft<T>`. In order to do this you may need to add new error variants:

https://github.com/pingcap/raft-rs/blob/18f1cd8bb1a452dd9a2e3a8c0b507fc138c1862d/src/errors.rs#L19-L22.

### PR DESCRIPTION
**Pull request by kleimkuhler**
This allows a `ConfigInvalid` error to propogate up and let user handle it.

Fixes #94

### CODE REVIEW

**Comment by DorianZheng** [comment]
@Hoverbear @BusyJay PTAL

File: src/raft.rs [hunk 1]
```diff
@@ -223,8 +223,8 @@ pub fn quorum(total: usize) -> usize {
 
 impl<T: Storage> Raft<T> {
     /// Creates a new raft for use on the node.
-    pub fn new(c: &Config, store: T) -> Raft<T> {
-        c.validate().expect("configuration is invalid");
+    pub fn new(c: &Config, store: T) -> Result<Raft<T>> {
+        c.validate()?;
         let rs = store.initial_state().expect("");
```
Comment (Hoverbear) [inline src/raft.rs]: Maybe we can also push up this expect?

Comment (kleimkuhler) [reply src/raft.rs]: My intention here was to propagate the `Error:ConfigInvalid` returned by `validate()`. https://github.com/pingcap/raft-rs/blob/13ec5ac33a3800f89a6946cb8c54ceba4b84ab98/src/config.rs#L170

I can make the test case more explicit to show that? Changing my test case to the below would show that it propagates back up to the user.
```
fn test_new_raft_with_bad_config_errors() {
    let invalid_config = new_test_config(INVALID_ID, vec![1, 2], 1, 1);
    let raft = Raft::new(&invalid_config, new_storage());
    assert_eq!(
        raft.err(),
        Some(Error::ConfigInvalid("invalid node id".to_owned()))
    )
}
```

Comment (Hoverbear) [reply src/raft.rs]: Oh, yes, that's fine. I meant the `store.initial_state()` could probably use a `?` as well instead of an `expect()`.

Comment (kleimkuhler) [reply src/raft.rs]: Oh! Yes, sorry about that; I'll change that as well.

**Comment by Hoverbear** [comment]
@kleimkuhler  Whoops! We both forgot the benchmarks! Could you please bump those too?

**Comment by kleimkuhler** [comment]
@Hoverbear Totally missed the benchmarking; sorry about that! Everything should be good to go now.

**Comment by hicqu** [comment]
LGTM.

**Comment by Hoverbear** [comment]
Thanks a lot for your work on this @kleimkuhler !

**Pull request merged**